In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
import os
from google.colab import drive

In [3]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data_path = '/content/drive/MyDrive/SU Works/CPSC_5310_Project/saved_data/full_dataset.parquet'

In [5]:
print(f"Loading from {data_path}...")
df = pd.read_parquet(data_path)
print(f"Loaded {len(df):,} rows.")

Loading from /content/drive/MyDrive/SU Works/CPSC_5310_Project/saved_data/full_dataset.parquet...
Loaded 47,248,845 rows.


In [6]:
print(f"Dataset Shape: {df.shape}")
print("\n--- Column Types and Non-Null Counts ---")
df.info()

Dataset Shape: (47248845, 19)

--- Column Types and Non-Null Counts ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47248845 entries, 0 to 47248844
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int8          
 1   tpep_pickup_datetime   datetime64[ms]
 2   tpep_dropoff_datetime  datetime64[ms]
 3   passenger_count        int8          
 4   trip_distance          float32       
 5   pickup_longitude       float32       
 6   pickup_latitude        float32       
 7   RateCodeID             int8          
 8   store_and_fwd_flag     category      
 9   dropoff_longitude      float32       
 10  dropoff_latitude       float32       
 11  payment_type           int8          
 12  fare_amount            float32       
 13  extra                  float32       
 14  mta_tax                float32       
 15  tip_amount             float32       
 16  tolls_amount           float32    

In [7]:
display(df.head())

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1         1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2         1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3         1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4         1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude  RateCodeID  \
0           1.59        -73.993896        40.750111           1   
1           3.30        -74.001648        40.724243           1   
2           1.80        -73.963341        40.802788           1   
3           0.50        -74.009087        40.713818           1   
4           3.00        -73.971176        40.762428           1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -73.974785         40.750618             1   
1                  N         -73.994415         40.759109             1   
2                  N         -73.951820         40.824413             2   
3                  N         -74.004326         40.719986             2   
4                  N         -74.004181         40.742653             2   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0         12.0    1.0      0.5        3.25           0.0   
1         14.5    0.5      0.5        2.00           0.0   
2          9.5    0.5      0.5        0.00           0.0   
3          3.5    0.5      0.5        0.00           0.0   
4         15.0    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  
0                    0.3     17.049999  
1                    0.3     17.799999  
2                    0.3     10.800000  
3                    0.3      4.800000  
4                    0.3     16.299999

Dataset Dictionary & Feature Descriptions

**Total Records:** ~47 Million Rows (Combined)  

###  **Trip Metadata**
| Field Name | Type | Description | Key / Notes |
| :--- | :--- | :--- | :--- |
| **VendorID** | Categorical | TPEP provider that provided the record. | `1` = Creative Mobile Technologies<br>`2` = VeriFone Inc. |
| **tpep_pickup_datetime** | DateTime | The date and time when the meter was engaged. | Start of the trip. |
| **tpep_dropoff_datetime** | DateTime | The date and time when the meter was disengaged. | End of the trip. |
| **store_and_fwd_flag** | Categorical | Whether the trip record was held in vehicle memory before sending to the vendor. | `Y` = store and forward<br>`N` = not a store and forward trip |
| **RateCodeID** | Categorical | The final rate code in effect at the end of the trip. | `1`= Standard rate, `2`=JFK, `3`=Newark<br>`4`=Nassau/Westchester, `5`=Negotiated, `6`=Group ride |

###  **Location & Movement**
| Field Name | Type | Description | Notes |
| :--- | :--- | :--- | :--- |
| **passenger_count** | Integer | The number of passengers in the vehicle. | **Note:** This is a driver-entered value. |
| **trip_distance** | Float | The elapsed trip distance in miles reported by the taximeter. | |
| **pickup_longitude** | Float | Longitude where the meter was engaged. | |
| **pickup_latitude** | Float | Latitude where the meter was engaged. | |
| **dropoff_longitude** | Float | Longitude where the meter was disengaged. | |
| **dropoff_latitude** | Float | Latitude where the meter was disengaged. | |

###  **Financials & Payment**
| Field Name | Type | Description | Key / Notes |
| :--- | :--- | :--- | :--- |
| **payment_type** | Categorical | How the passenger paid for the trip. | `1`= Credit card, `2`= Cash, `3`= No charge<br>`4`= Dispute, `5`= Unknown, `6`= Voided trip |
| **fare_amount** | Float | The time-and-distance fare calculated by the meter. | Does not include surcharges or tax. |
| **extra** | Float | Miscellaneous extras and surcharges. | Usually \$0.50 or \$1.00 for rush hour/overnight. |
| **mta_tax** | Float | \$0.50 MTA tax automatically triggered based on the metered rate. | |
| **improvement_surcharge** | Float | \$0.30 improvement surcharge assessed at flag drop. | |
| **tip_amount** | Float | Tip amount. | **Important:** Automatically populated for **Credit Cards only**. Cash tips are not included (usually 0). |
| **tolls_amount** | Float | Total amount of all tolls paid in trip. | |
| **total_amount** | Float | The total amount charged to passengers. | Does not include cash tips. |

In [8]:
print("\nMissing Values:")
print(df.isnull().sum())


Missing Values:
VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
pickup_longitude         0
pickup_latitude          0
RateCodeID               0
store_and_fwd_flag       0
dropoff_longitude        0
dropoff_latitude         0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    3
total_amount             0
dtype: int64


3 missing values are found in the `improvement_surcharge` feature

In [9]:
df = df.sort_values(by='tpep_pickup_datetime')

In [10]:
df.describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  4.724884e+07                    47248845                    47248845   
mean   1.529570e+00  2015-11-01 19:34:22.551000  2015-11-01 19:49:31.111000   
min    1.000000e+00         2015-01-01 00:00:00         2015-01-01 00:00:00   
25%    1.000000e+00         2015-01-30 07:34:06         2015-01-30 07:47:19   
50%    2.000000e+00         2016-01-31 21:20:27         2016-01-31 21:35:50   
75%    2.000000e+00         2016-03-02 05:54:00         2016-03-02 06:08:11   
max    2.000000e+00         2016-03-31 23:59:59         2016-06-29 15:58:16   
std    4.991248e-01                         NaN                         NaN   

       passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count     4.724884e+07   4.724884e+07      4.724884e+07     4.724884e+07   
mean      1.667040e+00   7.508401e+00     -7.276459e+01     4.008460e+01   
min       0.000000e+00  -3.390584e+06     -1.616987e+02    -7.703949e+01   
25%       1.000000e+00   1.000000e+00     -7.399166e+01     4.073611e+01   
50%       1.000000e+00   1.690000e+00     -7.398155e+01     4.075335e+01   
75%       2.000000e+00   3.090000e+00     -7.396657e+01     4.076779e+01   
max       9.000000e+00   1.907263e+07      9.464387e+01     4.047000e+02   
std       1.322092e+00   6.487658e+03      9.381830e+00     5.168528e+00   

         RateCodeID  dropoff_longitude  dropoff_latitude  payment_type  \
count  4.724884e+07       4.724884e+07      4.724884e+07  4.724884e+07   
mean   1.038543e+00      -7.282501e+01      4.011889e+01  1.352482e+00   
min    1.000000e+00      -7.401667e+02     -7.703949e+01  1.000000e+00   
25%    1.000000e+00      -7.399120e+01      4.073459e+01  1.000000e+00   
50%    1.000000e+00      -7.397962e+01      4.075380e+01  1.000000e+00   
75%    1.000000e+00      -7.396225e+01      4.076920e+01  2.000000e+00   
max    9.900000e+01       8.527402e+01      4.595333e+02  5.000000e+00   
std    5.902423e-01       9.150215e+00      5.040345e+00  4.922387e-01   

        fare_amount         extra       mta_tax    tip_amount  tolls_amount  \
count  4.724884e+07  4.724884e+07  4.724884e+07  4.724884e+07  4.724884e+07   
mean   1.239218e+01  3.243182e-01  4.977086e-01  1.794567e+00  2.843674e-01   
min   -9.576000e+02 -7.900000e+01 -1.000000e+00 -2.208000e+02 -9.999000e+01   
25%    6.500000e+00  0.000000e+00  5.000000e-01  0.000000e+00  0.000000e+00   
50%    9.000000e+00  0.000000e+00  5.000000e-01  1.250000e+00  0.000000e+00   
75%    1.400000e+01  5.000000e-01  5.000000e-01  2.260000e+00  0.000000e+00   
max    4.294967e+05  9.999900e+02  8.970000e+01  3.950589e+06  1.450090e+03   
std    7.861771e+01  5.158948e-01  4.672388e-02  5.747384e+02  1.657184e+00   

       improvement_surcharge  total_amount  
count           4.724884e+07  4.724884e+07  
mean            2.952470e-01  1.559273e+01  
min            -3.000000e-01 -9.584000e+02  
25%             3.000000e-01  8.300000e+00  
50%             3.000000e-01  1.160000e+01  
75%             3.000000e-01  1.715000e+01  
max             3.000000e-01  3.950612e+06  
std             3.812080e-02  5.801393e+02

## 1. Spatial Anomalies
The coordinate data contains extreme outliers that fall outside the bounds of New York City and, in some cases, the Earth's valid latitude range.
* Latitude Errors:
  * `pickup_latitude` Max: 404.7 (Valid range is -90 to +90).
  * `pickup_latitude` Min: -77.03 (Antarctica).
* Longitude Errors:
  * `pickup_longitude` Min: -161.69 (Near Hawaii/Alaska).
  * `pickup_longitude` Max: 94.64 (Eastern Hemisphere).

Similar outliers/noise are seen for the dropoff coordinates as well.

Impact: These outliers will severely distort modeling and must be filtered to a strictly defined NYC bounding box.

## 2. Metric Anomalies (Distance & Passenger Counts)
* Negative Distance: `trip_distance` Min is -3,390,584. It is physically impossible to travel negative distance.
* Extreme Distance: trip_distance Max is 1.90 x 10⁷ (approx. 19 million miles). This is clearly sensor noise or data corruption.
* Zero Passengers: passenger_count Min is 0. While technically possible (package delivery), these records may not reflect human transportation demand.
## 3. Financial Anomalies (Fares & Totals)
* Negative Fares: `fare_amount` and `total_amount` have minimums around -\$958. This likely indicates chargebacks, disputes, or voided transactions.
* Extreme Outliers: `total_amount` Max is ~\$3.95 Million. This is a data entry error that will destabilize the Loss Function (MSE) during model training.

In [11]:
df['trip_duration_min'] = (df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60

In [12]:
# Adding 0.001 to hours to prevent "Division by Zero" errors for 0-second trips
df['speed_mph'] = df['trip_distance'] / ((df['trip_duration_min'] / 60) + 0.001)

In [13]:
df[['trip_duration_min', 'speed_mph']].describe()

trip_duration_min     speed_mph
count       4.724884e+07  4.724884e+07
mean        1.514268e+01  9.278157e+01
std         3.790111e+02  1.759893e+05
min        -5.371001e+05 -3.335000e+08
25%         6.316667e+00  7.909604e+00
50%         1.045000e+01  1.057881e+01
75%         1.678333e+01  1.430052e+01
max         5.485556e+05  5.953275e+08

## 4. Outlier Observations for New Features

The descriptive statistics for `trip_duration_min` and `speed_mph` reveal significant anomalies:

*   **`trip_duration_min`**:
    *   **Min**: -537,100 minutes (physically impossible negative duration).
    *   **Max**: 548,555 minutes (approximately 380 days, an extremely long and unlikely trip duration).

*   **`speed_mph`**:
    *   **Min**: -333 million MPH (physically impossible negative speed).
    *   **Max**: 595 million MPH (astronomically high speed, far exceeding any real-world vehicle capabilities).

These extreme values highlight severe data quality issues. These outliers will distort exploratory analyses and models. We need robust cleaning and outlier handling strategies.